rfcs/001-alternate-crypto-backends.md

```diff
@@ -0,0 +1,100 @@
+# RFC001: Alternative crypto backends
+
+Issue: rustls/rustls#521
+
+## Why
+
+There are a bunch of reasons someone might want use crypto implementations
+other than those from *ring*.  The issue contains a thorough discussion,
+but to summarise:
+
+- runtime platform support: especially platforms like wasm who have their own
+  cryptography APIs.
+- CPU architecture support: we should have a support route for all architectures
+  that rustc can target.
+- upstream packaging requirements: RHEL has a hard policy that all cryptography
+  must come from a limited set of packages.
+- hardware support: including HSMs and cryptographic acceleration IP on microcontrollers.
+- assorted "pure rust TLS stack" reasons: including build system, memory safety, portability, etc.
+
+## Study
+
+First, let's look at where we use crypto and for what.  This includes
+uses which _already support_ alternative crypto backends via existing
+extension points.
+
+| Purpose | Current status | Example algorithm | Example code location |
+| --- | --- | --- | --- |
+| Random material generation | Statically uses *ring* | System-provided RNG | `src/rand.rs` |
+| Bulk encryption/decryption | Statically uses *ring* | AES-128-GCM | `src/cipher.rs` |
+| Certificate and signature verification | Pluggable extension point, behind `dangerous_configuration`. Default uses webpki which statically uses *ring*. | ECDSA-P256 | `src/verify.rs` |
+| Authentication key loading and signature generation | Pluggable extension point.  Default uses *ring*. | ECDSA-P256 | `src/sign.rs` |
+| Key exchange | Statically uses *ring* | X25519 | `src/kx.rs` |
+| TLS1.2 PRF | Statically uses *ring* | HMAC-SHA256 | `src/tls12/kdf.rs` |
+| TLS1.3 key schedule | Statically uses *ring* | HKDF-HMAC-SHA256 | `src/tls13/key_schedule.rs` |
+| Server ticket production | Pluggable extension point. Default does not support ticket generation. Provided optional implementation uses *ring*. | CHACHA20-POLY1305 | `src/ticketer.rs` |
```

Comment:
Another often-forgotten fact is that most hardware-accelerated cryptography is asynchronous.  Hardware accelerators have very deep pipelines with substantial latency, so one must have a large number of requests in flight to get good performance.  In the case of rustls, this means continuing to receive data and enqueue decryption requests before knowing if the earlier decryption requests succeeded.  Hardware accelerators are often many times faster than the CPU clusters they are attached to and use much less power, so it is worthwhile to support them if possible.